neo/Cryptography/ECC/ECDsa.cs

```diff
@@ -104,5 +106,55 @@ public bool VerifySignature(byte[] message, BigInteger r, BigInteger s)
             BigInteger v = point.X.Value.Mod(curve.N);
             return v.Equals(r);
         }
+
+        public static ECPoint KeyRecover(ECCurve curve, BigInteger r, BigInteger s, byte[] msg, bool isEven, bool doChecks)
+        {
+            if (r < BigInteger.One || s < BigInteger.One)
+            {
+                throw new ArithmeticException("Invalid signature");
+            }
+            // calculate h
+            BigInteger h = (curve.Q + 1 + 2 * (BigInteger)Math.Sqrt((double)curve.Q)) / curve.N;
+            BigInteger e;
+            ECPoint Q = new ECPoint();
+            int messageBitLength;
+
+            for (int i = 0; i <= h; i++)
+            {
+                // step 1.1 x = (n * i) + r
+                BigInteger Rx = curve.N * i + r;
+                if (Rx > curve.Q) break;
+
+                // step 1.2 and 1.3 get point R
+                ECPoint R;
+                if (isEven)
+                {
+                    R = ECPoint.DecompressPoint(0, Rx, curve);
+                }
+                else
+                {
+                    R = ECPoint.DecompressPoint(1, Rx, curve);
+                }
+                if (doChecks)
```

Comment:
I think that we should always do the checks

Reply:
Fixed